src/shims/unix/fs.rs

```diff
@@ -127,6 +129,89 @@ impl FileDescription for FileHandle {
         }
     }
 
+    fn flock<'tcx>(
+        &self,
+        communicate_allowed: bool,
+        op: FlockOp,
+    ) -> InterpResult<'tcx, io::Result<()>> {
+        assert!(communicate_allowed, "isolation should have prevented even opening a file");
+        #[cfg(target_family = "unix")]
+        {
+            use std::os::fd::AsRawFd;
+
+            // Transform `op` into host's flags
+            use FlockOp::*;
+            let host_op = match op {
+                SharedLock { nonblocking: false } => libc::LOCK_SH,
+                SharedLock { nonblocking: true } => libc::LOCK_SH | libc::LOCK_NB,
+                ExclusiveLock { nonblocking: false } => libc::LOCK_EX,
+                ExclusiveLock { nonblocking: true } => libc::LOCK_EX | libc::LOCK_NB,
+                Unlock => libc::LOCK_UN,
+            };
+
+            let fd = self.file.as_raw_fd();
+            let ret = unsafe { libc::flock(fd, host_op) };
```

Comment:
in the blocking modes this will just block the interpreter, even though it may have other interpreted threads that it could make progress on. I think we'll need to do some emulation of blocking mode by using the nonblocking operations and blocking the interpreter thread only

Reply:
Ouch, yeah, good catch -- we can't call blocking host operations in Miri. We shouldn't expose blocking `flock` at all for now; some significant design work is needed before we can do this. See also https://github.com/rust-lang/miri/issues/3449.

Reply:
Maybe we could use non-blocking `flock` even when blocking was requested and raise an error in case if it would block? I think it should be a good enough workaround for the time being.

Reply:
We already support putting threads to sleep. I think this could be a good candidate for implementing #3449 without having to go all the way to arbitrary socket support.

Reply:
So you suggest to use non-blocking `flock` and if it fails put thread to sleep? Can you point to a relevant piece of code, which I could use as a reference?

Reply:
> Maybe we could use non-blocking `flock` even when blocking was requested and raise an error in case if it would block? I think it should be a good enough workaround for the time being.

that may be the better immediate step to land this PR, then the "put thread to sleep" logic can be implemented in a separate PR as there's a bit of design discussions that have to happen first

Reply:
I also think it'd be better to just do non-blocking in this PR, and leave blocking for later.

> Can you point to a relevant piece of code, which I could use as a reference?

That's the thing, there's no code like that in Miri yet, we don't support any operations that block on the host. All blocking so far is entirely controlled by Miri itself (like blocking on a futex that is managed by Miri).

Reply:
>I also think it'd be better to just do non-blocking in this PR, and leave blocking for later.

I would prefer to have a rudimentary blocking `flock` support. How about my suggestion above to throw an error when blocking `flock` is requested, but non-blocking `flock` has failed?

Reply:
> Maybe we could use non-blocking flock even when blocking was requested and raise an error in case if it would block? I think it should be a good enough workaround for the time being.

Yes that sounds good, that's pretty much what I meant by not supporting blocking for now.